### PR TITLE
Caching section

### LINKED ISSUE #44: Add a section with info about Caching

### PR DESCRIPTION
**Pull request by Borales**
Fixes #44

### CODE REVIEW

**Comment by cane-lagerwaard** [comment]
@Borales perhaps we should mention something about that it's using "LRU" (Last Resource Used) https://www.apollographql.com/docs/apollo-server/features/data-sources#using-memcachedredis-as-a-cache-storage-backend maybe also link to apollo docs as reference?

**Comment by Borales** [comment]
@cane-lagerwaard added

File: docs/falcon-server/caching.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+`query` - is a name of the GQL operation and `menu` is a name of the Query field. So in this case, instead of using a predefined TTL,
+Falcon-Server will cache the response of `query.menu` resolver for 30 minutes (from the config above).
+
+> `@cache` directive works on a Schema level only, meaning that you cannot mark existing Queries as cacheable by yourself
+> via config to avoid introducing unintended behavior of the Extension.
+
+### Overriding cache options
+
+Cache options are being checked with this order (from higher priority to the lowest):
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
Cache options are being checked with this order (from highest to lowest priority):
```

File: docs/falcon-server/caching.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+---
+title: Caching
+---
+
+Falcon-Server exposes `@cache` directive through the Base GraphQL Schema to the rest of the extensions to define
+their own caching options. Such cache will be handled by `@cache` directive itself, so the Extension does not have
+to do anything by itself in terms of cache checks, cache-key generations etc.
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
to do anything by itself in terms of cache checks, cache-key generations, etc.
```

File: docs/falcon-server/caching.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+## Schema Caching options
+
+To use cache within your Extension - you need to put `@cache` directive next to your field you want to cache
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
To use cache within your Extension - you need to put `@cache` directive next to the field you want to cache
```

File: docs/falcon-server/caching.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+in Extension's GraphQL Schema:
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
in the Extension's GraphQL Schema:
```

File: docs/falcon-server/caching.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+```graphql
+type Query {
+  myData: Data @cache
+  myOtherData: Data @cache(ttl: 20)
+}
+```
+
+In this case - `myData` Query will be cached by Falcon-Server with a default cache TTL, and `myOtherData` Query
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
In this example - `myData` Query will be cached by Falcon-Server with a default cache TTL, and `myOtherData` Query
```

File: docs/falcon-server/caching.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+will be cached with a cache TTL of `20` (20 minutes).
+
+> Default cache TTL equals `10` (10 minutes)
+
+## Sub-caching
+
+It is also possible to cache a nested data of your Extension (which for example may require some heavy requests
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
It is also possible to cache nested data of your Extension (which for example may require some heavy requests
```

File: docs/falcon-server/caching.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  data(id: String): Data @cache
+}
+
+type Data {
+  nestedData: [String] @cache
+}
+```
+
+In order to ensure the uniqueness of the cache key - `@cache` directive takes into account the name
+of the cached field, all input arguments for the current resolver and also checks its the "parent" value:
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
of the cached field, all input arguments for the current resolver and also checks its "parent" value:
```

File: docs/falcon-server/caching.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    "default": {
+      "ttl": 60
+    }
+  }
+}
+```
+
+> You could use special values for your `production` environment via `config/production.json` config file
+
+Every author of the Extension can "mark" his Queries as "cacheable" by putting the `@cache` directive next to the corresponding
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
Every author of the Extension can "mark" their Queries as "cacheable" by putting the `@cache` directive next to the corresponding
```

File: docs/falcon-server/caching.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+queries in his Extension's GQL Schema. In some cases - you may disagree with his decisions and you may want to use
```
Comment (cane-lagerwaard) [inline docs/falcon-server/caching.md]:
```suggestion
queries in his Extension's GQL Schema. In some cases - you may disagree with these decisions and you may want to use
```

File: docs/falcon-server/caching.md (hunk 8 above)
Comment (Borales) [reply docs/falcon-server/caching.md]: @cane-lagerwaard what about "**provided** Queries"?

File: docs/falcon-server/falcon-server-api.md [hunk 10]
```diff
@@ -88,7 +88,26 @@ It should provide a simple logic to determine a custom priority or return a defa
 This method must be defined if a specific ApiDataSource supports Dynamic Routing.
 It will be called whenever Dynamic Route chooses this ApiDataSource instance to resolve URL type
 for the provided `path`. See Dynamic Route result response type
-[here](/docs/falcon-server/basics#dynamicrouteresolver-result-structure).
+[here](falcon-server/basics.md#dynamicrouteresolver-result-structure).
+
+#### `api.getCacheContext()`
+
+This is an optional method to define Cache Context object for your ApiDataSource instance.
+If your API does not require any distinguish data to be checked while calculating a cache key -
```
Comment (cane-lagerwaard) [inline docs/falcon-server/falcon-server-api.md]:
```suggestion
If your API does not require any distinctive data to be checked while calculating a cache key -
```

**Pull request merged**